backend/wayland/backend.c

```diff
@@ -421,49 +436,69 @@ struct wlr_backend *wlr_wl_backend_create(struct wl_display *display,
 	}
 	wl_event_source_check(wl->remote_display_src);
 
-	wlr_log(WLR_DEBUG, "Opening DRM render node %s", wl->drm_render_name);
-	wl->drm_fd = open(wl->drm_render_name, O_RDWR | O_NONBLOCK | O_CLOEXEC);
-	if (wl->drm_fd < 0) {
-		wlr_log_errno(WLR_ERROR, "Failed to open DRM render node %s",
-			wl->drm_render_name);
-		goto error_remote_display_src;
-	}
+	wl->drm_fd = -1;
+	if (wl->drm_render_name != NULL) {
+		wlr_log(WLR_DEBUG, "Opening DRM render node %s", wl->drm_render_name);
+		wl->drm_fd = open(wl->drm_render_name, O_RDWR | O_NONBLOCK | O_CLOEXEC);
+		if (wl->drm_fd < 0) {
+			wlr_log_errno(WLR_ERROR, "Failed to open DRM render node %s",
+				wl->drm_render_name);
+			goto error_remote_display_src;
+		}
 
-	int drm_fd = fcntl(wl->drm_fd, F_DUPFD_CLOEXEC, 0);
-	if (drm_fd < 0) {
-		wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
-		goto error_drm_fd;
-	}
+		int drm_fd = fcntl(wl->drm_fd, F_DUPFD_CLOEXEC, 0);
+		if (drm_fd < 0) {
+			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
+			goto error_drm_fd;
+		}
 
-	struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
-	if (gbm_alloc == NULL) {
-		wlr_log(WLR_ERROR, "Failed to create GBM allocator");
-		close(drm_fd);
-		goto error_drm_fd;
+		struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
+		if (gbm_alloc == NULL) {
+			wlr_log(WLR_ERROR, "Failed to create GBM allocator");
+			close(drm_fd);
+			goto error_drm_fd;
+		}
+		wl->allocator = &gbm_alloc->base;
+	} else {
+		wlr_log(WLR_DEBUG, "No render node found, falling back to shared memory");
+		struct wlr_shm_allocator *shm_alloc = wlr_shm_allocator_create();
+		if (shm_alloc == NULL) {
+			wlr_log(WLR_ERROR, "Failed to create shared memory allocator");
+			goto error_remote_display_src;
+		}
+		wl->allocator = &shm_alloc->base;
 	}
-	wl->allocator = &gbm_alloc->base;
 
 	wl->renderer = wlr_renderer_autocreate(&wl->backend);
 	if (wl->renderer == NULL) {
 		wlr_log(WLR_ERROR, "Failed to create renderer");
 		goto error_allocator;
 	}
 
+	const struct wlr_drm_format_set *remote_formats;
+	const struct wlr_drm_format_set *render_formats;
+	if (wl->drm_fd >= 0) {
+		remote_formats = &wl->linux_dmabuf_v1_formats;
+		render_formats = wlr_renderer_get_dmabuf_render_formats(wl->renderer);
+	} else {
+		remote_formats = &wl->shm_formats;
+		render_formats = NULL; // TODO
```

Comment:
Still not yet sure what the best way forward is here:

1. The renderer will use `get_data_ptr` to write to the buffers. It doesn't care whether the buffers are DMA-BUFs (DRM dumb buffers) or shm.
2. The backend doesn't care how the renderer writes to the buffers. It only needs to make sure that the renderer can write to its shm or DRM dumb buffers.

Maybe we can assume renderer will never have different format constraints for DRM dumb buffers and shm. I wonder if we'll ever have a case of a renderer which is able to write to both GPU buffers and CPU buffers, with a different set of formats for each case?

Potential solutions:

- Assume renderers will only ever support one buffer type for rendering. Rename `wlr_renderer_get_dmabuf_render_formats` to `wlr_renderer_get_render_formats`. Note, this interface is a little bit too permissive, since software renderers don't have modifiers, but the allocator interface has the same issue too.
- Add `wlr_renderer_get_software_render_formats` (or `wlr_renderer_get_shm_render_formats` for consistency with existing functions?), which returns a flat list of DRM formats, just like `wlr_renderer_get_shm_texture_formats`.

Reply:
I like the `wlr_renderer_get_shm_render_formats`, its more consistent with the existing functions. Maybe we could hide that behind an impl function and just call `wlr_renderer_get_render_formats`. Is the renderer aware if it's using dmabufs or not?

Reply:
> I like the wlr_renderer_get_shm_render_formats, its more consistent with the existing functions.

If we keep it private for now, it's also the safest option, since it doesn't change the existing API. We can always revisit this when we decide to make it public.

> Is the renderer aware if it's using dmabufs or not?

The renderer can't know in advance if the allocator will produce DMA-BUFs or something else.

Reply:
I implemented it in the pixman render PR in https://github.com/swaywm/wlroots/pull/2661/commits/090c8f6d6a7070fb31cc9aa3edf0a73ecdb68eeb, and use it in https://github.com/swaywm/wlroots/pull/2661/commits/d7571f25b09e8a3a6be932d7789996f0cfcb7d51. I chose to use the `wlr_drm_format_set` in case we merge the two functions to get the render formats.